// @angular-eslint/eslint-plugin/dist/rules/prefer-standalone.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RULE_NAME = void 0;
const utils_1 = require("@angular-eslint/utils");
const create_eslint_rule_1 = require("../utils/create-eslint-rule");
exports.RULE_NAME = 'prefer-standalone';
const METADATA_PROPERTY_NAME = 'standalone';
const IS_STANDALONE = 'true';
exports.default = (0, create_eslint_rule_1.createESLintRule)({
    name: exports.RULE_NAME,
    meta: {
        type: 'suggestion',
        docs: {
            description: `Ensures component, directive and pipe \`${METADATA_PROPERTY_NAME}\` property is set to \`${IS_STANDALONE}\` in the component decorator`,
        },
        fixable: 'code',
        schema: [],
        messages: {
            preferStandalone: `The {{type}} \`${METADATA_PROPERTY_NAME}\` property should be set to \`${IS_STANDALONE}\``,
        },
    },
    defaultOptions: [],
    create(context) {
        const standaloneRuleFactory = (type) => (node) => {
            const standalone = utils_1.ASTUtils.getDecoratorPropertyValue(node, METADATA_PROPERTY_NAME);
            if (standalone &&
                utils_1.ASTUtils.isLiteral(standalone) &&
                standalone.value === true) {
                return;
            }
            if (!utils_1.ASTUtils.getDecoratorArgument(node)) {
                return;
            }
            context.report({
                node: nodeToReport(node),
                messageId: 'preferStandalone',
                data: { type },
                fix: (fixer) => {
                    if (standalone &&
                        utils_1.ASTUtils.isLiteral(standalone) &&
                        standalone.value !== true) {
                        return [fixer.replaceText(standalone, IS_STANDALONE)].filter(utils_1.isNotNullOrUndefined);
                    }
                    return [
                        utils_1.RuleFixes.getDecoratorPropertyAddFix(node, fixer, `${METADATA_PROPERTY_NAME}: ${IS_STANDALONE}`),
                    ].filter(utils_1.isNotNullOrUndefined);
                },
            });
        };
        return {
            [utils_1.Selectors.COMPONENT_CLASS_DECORATOR]: standaloneRuleFactory('component'),
            [utils_1.Selectors.DIRECTIVE_CLASS_DECORATOR]: standaloneRuleFactory('directive'),
            [utils_1.Selectors.PIPE_CLASS_DECORATOR]: standaloneRuleFactory('pipe'),
        };
    },
});
function nodeToReport(node) {
    if (!utils_1.ASTUtils.isProperty(node)) {
        return node;
    }
    return utils_1.ASTUtils.isMemberExpression(node.value)
        ? node.value.property
        : node.value;
}
